lib/index.js

```diff
@@ -0,0 +1,233 @@
+    electors,
+    // optional `f+1` "trusted" electors that must not fail and can handle
+    // catastrophic case where remaining `2f` electors cannot communicate
+    // (recovery electors are not a failsafe against a coordinated byzantine
+    // attack as the byzantine electors could collude to show different sides
+    // of a fork, but safe against network failures preventing consensus)
+    recoveryElectors,
+    // number of generations of merge events where no new electors participate
+    // and fewer than 2f+1 total electors (including self) participate causing
+    // recovery mode to trigger because consensus cannot be reached
+    recoveryGenerationThreshold,
+    // number of recovery electors that must have seen a consensus decision
+    // prior to triggering recovery mode in order for the decision to be
+    // accepted, avoiding recovery mode; can also be thought of as the number
+    // of recovery electors an attacker must take over (and if the attacker
+    // can prevent the other recovery electors from communicating with the
+    // non-recovery electors) to cause a fork whereby recovery mode is both
+    // triggered and consensus is decided
+    recoveryDecisionThreshold
+  };
+};
+
+// it is useful to override this function in tests
+api._computeElectors = async ({
+  blockHeight, electorPoolElectors, latestBlockSummary, ledgerConfiguration,
+  ledgerNode
+}) => {
+  // get partipicants for the last block
+  const {_blocks} = ledgerNode.consensus;
+  const {consensusProofPeers, mergeEventPeers} = await _blocks.getParticipants(
+    {blockHeight: blockHeight - 1, ledgerNode});
+
+  // TODO: we should be able to easily remove previously detected
+  // byzantine nodes (e.g. those that forked at least) from the electors
+
+  // TODO: simply count consensus event signers once and proof signers
+  //   twice for now -- add comprehensive elector selection and
+  //   recommended elector vote aggregating algorithm in v2
+  const aggregate = {};
+  // add merge event peers
+  for(const id of mergeEventPeers) {
+    aggregate[id] = {id, weight: 1};
+  }
+
+  // TODO: weight previous electors more heavily to encourage continuity
+  // add consensusProof peers
+  for(const id of consensusProofPeers) {
+    if(id in aggregate) {
+      aggregate[id].weight = 3;
+    } else {
+      aggregate[id] = {id, weight: 2};
+    }
+  }
+  // add peers from the electorPool
+  for(const {id} of electorPoolElectors) {
+    // FIXME: weight these like this?
+    aggregate[id] = {id, weight: 4};
+  }
+
+  let electors = Object.values(aggregate);
+
+  // get elector count, defaulting to MAX_ELECTOR_COUNT if not set
+  // (hardcoded, all nodes must do the same thing -- but ideally this would
+  // *always* be set)
+
+  // FIXME: should the default `MAX_ELECTOR_COUNT` be removed given the
+  // requirements around recovery mode?  Specifically, *all* recovery mode mode
+  // electors (whatever number) *must* be electors as well.
```

Comment:
This is in reference to this line: https://github.com/veres-one/veres-one-consensus-continuity-elector-selection/blob/b2d01abd21efb09a246797e5c9ebdfa832e67987/lib/index.js#L134-L135

MAX_ELECTOR_COUNT is currently set to 10.  Should this limitation be removed entirely, or should it be something like:
```
const electorCount = electorPoolElectors.length + (ledgerConfiguration.electorCount || MAX_ELECTOR_COUNT);
```

Reply:
I think the math there is wrong, but I'm fine with removing it completely and requiring it to be set in the ledger config.

Reply:
If we do that, is there any possibility that a mismatch between the `ledgerConfiguration.electorCount` and the number of electors defined in the electorPool document could result in a broken ledger.  For instance....

1. `ledgerConfiguration.electorCount` = 5;
2. The electorPool electors grows from 5 to 10.
3. We have a requirement that ALL recovery electors are electors.
4. The ledger attempts to go into recovery mode.
5. Not all recovery electors are electors.
6. Weep.

Reply:
I think I am interpreting "We have a requirement that ALL recovery electors are electors" differently from you. You seem to be meaning this:

"We have a requirement that ALL recovery electors are electors *FOR EVERY BLOCK*" when I am not. I am just saying that you can't be a recovery elector AND also NOT be a regular elector for a given block. That makes no statement about whether you are chosen to be an elector for a particular block. It is just about your capabilities SHOULD you be chosen.

So, to try and clarify things. Let's look at this from a functional/algorithmic perspective. Let's say that the goal is to pick electors for block X.

1. Read the ledger configuration.
2. Read the elector pool.
3. Compute the number of electors to use:
3.1. `electorCount = Math.min(electorPool.size, config.maxElectorCount)`
3.2. Decrease `electorCount` to match a form of `3f+1`.
4. Compute the number of recovery mode electors:
4.1. `recoveryElectorCount = f+1`.
5. Create an empty set of `electors` for block X.
6. Create an empty set of `recoveryElectors` for block X.
6. While we haven't picked `electorCount` electors:
6.1. If are guarantor electors that haven't been picked yet, deterministically pick one and add it to `electors`. Note: guarantor electors must also be recovery mode electors, so add them to `recoveryElectors` if `recoveryElectors` is not yet full. Continue (loop early).
6.2. If there aren't enough recovery mode electors yet, deterministically pick one and add it to `electors` and `recoveryElectors` and continue (loop early).
6.3. Deterministically pick an elector from the elector pool and add it to `electors`.
7. Return `electors` and `recoveryElectors`.

Reply:
> 5. Not all recovery electors are electors.

I'm not really sure what you mean here because this can't ever happen. You can't be a recovery elector that doesn't also function as an elector (it is an error for the recovery electors to not be a subset of the electors). But there's further confusion for me because this happens after the ledger has gone into recovery mode in your step 4. So I assume you're talking about the *new* electors that are chosen ... and here there are no recovery electors at that point at all... so I don't know why you're talking about them.